app/src/main/java/org/onionshare/android/tor/TorManager.kt

```diff
@@ -70,175 +41,107 @@ private const val MOAT_FRONT = "ajax.aspnetcdn.com"
-                val progress = (percent * 0.7).roundToInt()
-                changeStartingState(progress)
-            }
-            return@RawEventListener
-        }
-        val onion = (state.value as? TorState.Starting)?.onion
-        // descriptor upload counts as 90%
-        if (state.value !is TorState.Started) {
-            if (onion != null && keyword == EVENT_HS_DESC && data.startsWith("UPLOAD $onion")) {
-                changeStartingState(90, onion)
-            }
-        }
-        // We consider already the first upload of the onion descriptor as started (100%).
-        // In practice, more uploads are needed for the onion service to be reachable.
-        if (onion != null && keyword == EVENT_HS_DESC && data.startsWith("UPLOADED $onion")) {
-            onDescriptorUploaded(onion)
-        }
-    }
-
     /**
-     * Starts [TorService] and creates a new onion service.
+     * Starts [ShareService] and creates a new onion service.
      * Suspends until the address of the onion service is available.
      */
-    @Suppress("BlockingMethodInNonBlockingContext")
     suspend fun start() = withContext(Dispatchers.IO) {
         if (state.value !is TorState.Stopped) stop()
 
         LOG.info("Starting...")
         val now = System.currentTimeMillis()
         _state.value = TorState.Starting(progress = 0, startTime = now, lastProgressTime = now)
-        startLatch = CountDownLatch(1)
-        app.registerReceiver(broadcastReceiver, IntentFilter(ACTION_STATUS))
-        app.registerReceiver(errorReceiver, IntentFilter(ACTION_ERROR))
-        broadcastReceiverRegistered = true
 
         Intent(app, ShareService::class.java).also { intent ->
             startForegroundService(app, intent)
         }
-        try {
-            executableManager.installObfs4Executable()
-            executableManager.installSnowflakeExecutable()
-            startLatch?.await() ?: error("startLatch was null")
-            startLatch = null
-            onTorServiceStarted()
-        } catch (e: Exception) {
-            LOG.warn("Error while starting Tor: ", e)
-            stop()
-        }
+
+        val torWrapper = TorModule.createTorWrapper(app, wakeLockManager)
+        tor = torWrapper
+        torWrapper.setObserver(this@TorManager)
+        torWrapper.start()
+        torWrapper.enableNetwork(true)
     }
 
     fun stop(failedToConnect: Boolean = false) {
         LOG.info("Stopping...")
         startCheckJob?.cancel()
         startCheckJob = null
         _state.value = TorState.Stopping(failedToConnect)
-        controlConnection = null
+        tor?.stop()
+        tor = null
         Intent(app, ShareService::class.java).also { intent ->
             app.stopService(intent)
         }
-        // wait for service to stop and broadcast when it gets destroyed
+    }
+
+    @OptIn(DelicateCoroutinesApi::class)
+    override fun onState(s: TorWrapper.TorState) {
+        when (s) {
+            STARTING_STOPPING -> onStopped()
```

Comment:
Will this cause spurious calls to onStopped() until we fix onionwrapper#3?

Reply:
The code currently says `if (!started || stopped) return STARTING_STOPPING;` So as soon as `started` is `true` I hope we shouldn't get this anymore, but you know this code better than me.

Reply:
I think that's right, but `started` becomes true after starting the Tor process. So I guess the question is whether onState() could ever be called before the Tor process finishes starting.

Reply:
Looks like this depends on the onionwrapper internal `updateState()` method. This may get called before Tor has started if it finds that it has already bootstrapped or built a circuit, though I have no idea how that could be possible. If it is possible, we'd need to implement https://code.briarproject.org/briar/onionwrapper/-/issues/3

Reply:
> This may get called before Tor has started if it finds that it has already bootstrapped or built a circuit, though I have no idea how that could be possible.

If Tor has enough cached state to build circuits then it's possible for it to bootstrap and build a circuit while we're still connecting and authenticating to the control port. In that case we won't receive control port events from Tor telling us that it's finished bootstrapping/is able to build a circut, so `TorWrapper#onStart()` checks those conditions as soon as the control port is opened.

So it seems that potentially `onState()` *could* be called with a `STARTING_STOPPING` state at startup, in which case we'd get a spurious call to `onStopped()` here.

Reply:
We are calling `onStopped()` only when `onState()` gives us `STOPPED`.